tools/discovery/main.go

```diff
@@ -0,0 +1,211 @@
+					return nil
+				}
+				// try to parse the network from the string.
+				// this will validate the network and return an error if it's invalid.
+				_, err := proxycmn.EigenDANetworkFromString(v)
+				return err
+			},
+		},
+		&cli.StringFlag{
+			Name:    directoryAddressFlagName,
+			Usage:   "EigenDADirectory contract address (overrides the default network address)",
+			EnvVars: []string{"EIGENDA_DIRECTORY_ADDRESS"},
+		},
+		&cli.StringFlag{
+			Name:    outputFormatFlagName,
+			Usage:   fmt.Sprintf("Output format. Must be one of: %v", validOutputFormats),
+			Value:   "table",
+			EnvVars: []string{"OUTPUT_FORMAT"},
+			Action: func(ctx *cli.Context, v string) error {
+				if !slices.Contains(validOutputFormats, strings.ToLower(v)) {
+					return fmt.Errorf("invalid output format: %s. Must be one of: %v", v, validOutputFormats)
+				}
+				return nil
+			},
+		},
+	}
+
+	app.Action = discoverAddresses
+	if err := app.Run(os.Args); err != nil {
+		log.Fatalf("application failed: %v", err)
+	}
+}
+
+func discoverAddresses(ctx *cli.Context) error {
+	outputFormat := strings.ToLower(ctx.String(outputFormatFlagName))
+	rpcURL := ctx.String(ethRpcUrlFlagName)
+
+	// Simple logging
+	logger := log.New(os.Stderr, "[discovery] ", log.LstdFlags)
+
+	// Connect to Ethereum
+	client, err := ethclient.Dial(rpcURL)
+	if err != nil {
+		return fmt.Errorf("dial Ethereum node at %s: %w", rpcURL, err)
+	}
+	logger.Printf("Connected to Ethereum node at %s", rpcURL)
+
+	directoryAddr := ctx.String(directoryAddressFlagName)
+	// cases:
+	// 1. directory address: use it directly
+	// 2. no directory address, but network provided: use the default directory address for the network
+	// 3. no directory address and no network: use the default network for the chainID
+	if directoryAddr == "" {
+		logger.Printf("No directory address provided, attempting to auto-detect.")
+		networkName := ctx.String(NetworkFlagName)
+		var network proxycmn.EigenDANetwork
+		if networkName != "" {
+			logger.Printf("network %s provided, attempting to use default directory address.", networkName)
+			network, err = proxycmn.EigenDANetworkFromString(ctx.String(NetworkFlagName))
+			if err != nil {
+				return err
+			}
+		} else {
+			logger.Printf("No network provided, attempting to auto-detect EigenDADirectory address from chain ID.")
+			chainID, err := client.ChainID(ctx.Context)
+			if err != nil {
+				return fmt.Errorf("chainID call: %w", err)
+			}
+			if chainID == nil {
+				return fmt.Errorf("received nil chainID from Ethereum client")
+			}
+			networks, err := proxycmn.EigenDANetworksFromChainID(chainID.String())
+			if err != nil {
+				return fmt.Errorf("DefaultEigenDANetworkFromChainID (chainid: %s): %w", chainID, err)
+			}
+			network = networks[0]
+			if len(networks) > 1 {
+				logger.Printf("Multiple EigenDA networks found for chain ID %s: %v. Using the first one: %s", chainID, networks, network)
```

Comment:
You claimed to agree with my comment in [this thread](https://github.com/Layr-Labs/eigenda/pull/1804#discussion_r2237800256)

But you're still simply choosing the first network in cases where there are multiple possible networks. I think we should just force users to explicitly choose the correct network in cases where there are multiple options, and throw an error here if they don't.

"Taking a guess" at the network a user wants doesn't seem helpful... what if we guess wrong? Maybe they'll see this log, but maybe not

Reply:
I was opting for having a simpler CLI, without forcing user to pass too many required flags before getting an answer. But I agree that choosing network is pretty fundamental. Refactored in [b882d9c](https://github.com/Layr-Labs/eigenda/pull/1804/commits/b882d9c443627de794235e20d354c545ba51ba9f) by making network flag required. This also greatly simplified the code. Good change!